Add exercise 3

otto-lightricks opened the pull request.

otto-lightricks committed:
Add exercise 3

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.h:
```diff
@@ -0,0 +1,27 @@
+
+@interface CardGameViewController : UIViewController
+
+@property (strong, nonatomic) Deck *deck;
```
In general it's better to make properties `readonly` whenever possible, especially for public properties like these.
Also, minor - you don't need the explicit `strong`, it's the default if you don't specify anything.

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.h:
```diff
@@ -0,0 +1,27 @@
+#import <UIKit/UIKit.h>
+
+@class CardMatchingGame, Deck;
+
```
Don't worry about it for this exercise, but we have a convention in Lightricks where every class and every public property/method has a `///` docstring describing what it is. 
In this case it might help to change the name of the class to indicate that it's intended as an abstract base class, as it's not really obvious just from looking at the code.

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.mm:
```diff
@@ -0,0 +1,97 @@
+  return _game;
+}
+
+- (CardMatchingGame *)initializeGame {
```
Name isn't quite right as it's creating a game, not initialising one.  `makeGame` might be better.

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.h:
```diff
@@ -0,0 +1,27 @@
+
+#import <UIKit/UIKit.h>
+
+@class CardMatchingGame, Deck;
```
There are a couple of obj-c macros - `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END`- that improve Obj-c compatibility with Swift (as then the compiler knows what is effectively optional and what isn't) and also improve the expressiveness of the code, so in general it's good practice to always encapsulate your code with this, both in headers and in implementation files.

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.mm:
```diff
@@ -0,0 +1,97 @@
+}
+
+- (Deck *)createDeck {
+  return nil;
```
If it's an "abstract" method where it would be a programmer error to call this implementation directly, you could consider putting an assertion in here.

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.mm:
```diff
@@ -0,0 +1,97 @@
+
+- (void)viewWillAppear:(BOOL)animated {
+  [super viewWillAppear:animated];
+  const auto appearance = [[UITabBarAppearance alloc] init];
```
Good to be explicit, but don't worry _too_ much about putting `const` everywhere.

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.mm:
```diff
@@ -0,0 +1,97 @@
+
+- (IBAction)modeChanged:(UISegmentedControl *)sender {
+  switch (sender.selectedSegmentIndex) {
+    case 0:
```
Could consider named constants for these cases?

joemuers replied:
or an enum?

joemuers commented on Matchismo/Matchismo/Controller/CardGameViewController.mm:
```diff
@@ -0,0 +1,97 @@
+}
+
+- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+  if ([segue.identifier isEqualToString: @"showHistory"]) {
```
Presuming this comes from the tutorial code that you're given, but just in case you're not aware we generally try to avoid xibs, storyboards and associated IB stuff at all costs. 
It's almost always better to do stuff in code.

joemuers commented on Matchismo/Matchismo/Controller/HistoryViewController.mm:
```diff
@@ -0,0 +1,46 @@
+
+- (void) setHistoryText:(NSAttributedString *)historyText {
+  _historyText = historyText;
+  if (self.view.window) [self updateUI];
```
Are we checking for the window just to check if we're attached to the view hierarchy?
Remember that just accessing `self.view` will as a side effect cause the view to be loaded.  Better to check `self.isViewLoaded` first.

joemuers replied:
Also we don't generally use implicit scopes like this in LT (in obj-c at least).  We prefer explicit `{` and `}` even for one-liners, and we usually include the newline. So:
```
if (self.isViewLoaded && self.view.window) {
  [self updateUI];
}
```

joemuers commented on Matchismo/Matchismo/Controller/HistoryViewController.h:
```diff
@@ -0,0 +1,14 @@
+
+@interface HistoryViewController : UIViewController
+
+@property (nonatomic) NSAttributedString *historyText;
```
Is it necessary to expose an `NSAttributedString`, as opposed to just an `NSString`?

otto-lightricks replied:
I used `NSAttributedString` since the VC doesn't need to know anything about different cards and how the card symbols should be coloured, it just displays the text that is given to it. I refactored the `NSAttributedString` properties out of the model classes into the Playing/Set Card VCs so they add the attributes to the correct ranges of the description string and those strings are stored in another `NSAttributedString` (so the attributes are not lost) that is then given to the HistoryVC when it is shown. 
Not sure if the way I handled the `descriptionLabel` and `moveHistory` in the card VCs in the best way or if there would've been a better way to handle how the VCs get the text from the `CardMatchingGame` model and display it and then provide the text to the HistoryVC.

joemuers replied:
Yeah it's a slightly tricky one. Obviously I'm not that familiar with the code, but I'm wondering if there's a slight problem in the design here in the sense that we have the `CardMatchingGame` object which vends a string describing some event (`lastMoveDescription`), that string is passed around and in other parts of the app we're having to do a regex search on it, add attributes and so on.  Might be easier if instead you express the "lastMove" as an actual model object (i.e. with bool/enum fields or whatever to indicate the card type, if it's a match... and so on), pass that around the app and only express it as a string as the point where it's needed for the UI.
Does that make sense?

If that's a big refactor and you want to just carry on with the exercise then don't worry too much making this kind of change, but maybe at least have a think about what could be done here.

joemuers replied:
By the way I've also just noticed some potential bugs here where we have an `NSMutableString` and we're assigning it directly to `NSString` properties/variables (for example in `AbstractCardGameViewController` at line 99).  The danger with doing this is that the `NSString` property is referencing something that underneath is still actually an `NSMutableString` instance, and therefore might be changed at any point.  This can cause problems because in code if you have an `NSString` property, you'll usually assume that it's immutable and won't change. 
```
NSMutableString *myMutableString = [@"abc" mutableCopy];
NSString *myImmutableString = myMutableString; // from this, we might expect myImmutableString to always be "abc"
[myMutableString appendString:@"def"]; // but after this line, myImmutableString will now be changed to "abcdef"
```

The usual way to deal with this is to always use copy when assigning to immutable types (NSString, NSArray, NSDictionary etc).  So in the example above we would have:
`NSString *myImmutableString = [myMutableString copy]; // this will now be property immutable`

joemuers commented on Matchismo/Matchismo/Controller/PlayingCardGameViewController.mm:
```diff
@@ -0,0 +1,89 @@
+}
+
+- (void)startNewGame {
+  if (![self.game startNewGameWithCardCount:self.cardsCollection.count usingDeck
```
Slightly weird newline placement here makes it a bit confusing to read.
Would be better like this:
```
  if (![self.game startNewGameWithCardCount:self.cardsCollection.count
                                  usingDeck:[self createDeck]]) {
    return;
  }
  ```

joemuers commented on Matchismo/Matchismo/Model/SetCardDeck.h:
```diff
@@ -0,0 +1,14 @@
+
+@interface SetCardDeck : Deck
+
+- (instancetype)initWithShapes:(NSArray *)shapes Colors:(NSArray *)colors;
```
Wrong capitalisation - `colors:(NSArray *)colors`

joemuers commented on Matchismo/Matchismo.xcodeproj/project.pbxproj:
```diff
@@ -94,8 +109,14 @@
 			children = (
-				27A76E652850AA79003BA2A1 /* ViewController.h */,
-				27A76E662850AA79003BA2A1 /* ViewController.mm */,
+				27A76E652850AA79003BA2A1 /* CardGameViewController.h */,
```
Doesn't matter here, but we keep files & folders alphabetised in the main code base (it's enforced).
There's a context menu entry in Xcode for doing this easily.

joemuers commented on Matchismo/Matchismo/Controller/PlayingCardGameViewController.mm:
```diff
@@ -0,0 +1,89 @@
+                                           :[self createDeck]]) {
+    return;
+  }
+  for (UIButton *cardButton in self.cardsCollection) {
```
Just to make it clearer what this block of code is doing, you could consider moving it to a new private method called `resetCardButtons` or something similar.

joemuers commented on Matchismo/Matchismo/Controller/SetCardGameViewController.mm:
```diff
@@ -0,0 +1,70 @@
+    [cardButton setEnabled:YES];
+    [cardButton setAlpha:1.0];
+    const auto cardButtonIndex = [self.cardsCollection indexOfObject:cardButton];
+    const auto *card = (SetCard *)[self.game cardAtIndex:cardButtonIndex];
```
Casting is pretty dangerous, and if you're having to cast something it's often a sign of some problem with the design. One of the problems with Objective-C is that this is very, very easy to do. 
If it's unavoidable you can make it safer by putting in an `isKindOfClass` check, but all the same it's preferable to fix the design if poss so that we can keep the typing completely static throughout.

joemuers commented on Matchismo/Matchismo/Model/Card.h:
```diff
@@ -14,7 +14,7 @@ struct MatchResult {
 };
 
-@property (strong, nonatomic) NSString *contents;
+@property (strong, nonatomic) NSAttributedString *contents;
```
Bit odd that we're using an attributed string in what's presumably a model object?
Better just to keep "business" logic here and put the UX-related logic (i.e. choice of fonts, colours or whatever) in the UI layer.

joemuers commented on Matchismo/Matchismo/Model/Card.h:
```diff
@@ -14,7 +14,7 @@ struct MatchResult {
   NSMutableArray<Card *> *matches;
```
Presumably this came from the provided code, but FYI using C-structs to hold Obj-C objects like this is pretty unusual, it's not something you'll see much in our codebase.

otto-lightricks replied:
I added that data structure myself to return two values from the function and be clear what types the two values are. Since objc doesn't have tuples is it normal to use an array or dictionary to return multiple values or is it done some other way? And how are different types of objects handled if an array is used, i.e. if the match function returned NSArray instead of MatchResult it isn't clear what the array contains so is it just assumed that the return type or what the array contains is known based on documentation and the objects are then cast to the correct type when used?

joemuers replied:
It's true that there's no native concept of tuples in ObjC, though there are some third party libraries that provide them (in LT we often use RACTuple from the Reactive Cocoa library).
The alternative though is just to define a simple class with a couple of properties.
Usage of structs in ObjC isn't really equivalent to usage of structs in Swift.  In general they're just used when you're dealing with simple C-types (e.g. CGRect, CGSize and so on). 

I'd also add that it's generally a bad idea to have mutable fields on model objects, data containers and so on.  If these kinds of objects are entirely readonly (i.e. immutable) then it's much easier to reason about the code and avoid bugs arising from shared mutable state. 

So in this case, I guess you just define an interface something like this:
```
@interface MatchResult : NSObject

@property (readonly, nonatomic) NSInteger score;
@property (readonly, nonatomic) NSArray<Card *> *matches;

- (id)init NS_UNAVAILABLE;
- (instancetype)initWithScore:(NSInteger)score andMatches:(NSArray<Card *> *)matches NS_DESIGNATED_INITIALISER;

@end
```

Regarding mixed types in an array and casting the elements - you totally _could_ do this and objective c would allow it, but it's a bad idea as you can't leverage any safety checking from the compiler, so it would be really easy to make a mistake and crash the whole app.

joemuers commented on Matchismo/Matchismo/Model/CardMatchingGame.h:
```diff
@@ -11,18 +11,19 @@
 @interface CardMatchingGame : NSObject
 
-enum GameMode : int {
+typedef enum {
```
Generally nicer to use the `NS_ENUM` macro to define Obj-C enums.

joemuers commented on Matchismo/Matchismo/Model/CardMatchingGame.h:
```diff
@@ -11,18 +11,19 @@
-@property (nonatomic, readonly) NSString *lastCardFlipDescription;
+@property (nonatomic) GameMode mode;
+@property (nonatomic, readonly) NSAttributedString *lastMoveDescription;
+@property (nonatomic) NSMutableAttributedString *moveHistory;
```
Exposing mutable stuff as a public property is generally a bad idea.  Is it possible to encapsulate this and hide the mutability?

joemuers replied:
Also, the fact that it's both mutable and settable (i.e. not `readonly`) doesn't quite make sense.

joemuers commented on Matchismo/Matchismo/Model/CardMatchingGame.mm:
```diff
@@ -12,23 +12,45 @@
 
 @implementation CardMatchingGame
 
+constexpr int TWO_MATCH_BONUS = 4;
```
Consider an enum as opposed to a bunch of free-floating constants?

joemuers commented on Matchismo/Matchismo/Model/CardMatchingGame.mm:
```diff
@@ -12,23 +12,45 @@
 
+- (void)setLastMoveDescription:(NSAttributedString *)lastMoveDescription {
+  _lastMoveDescription = lastMoveDescription;
+  if (![lastMoveDescription isEqualToAttributedString:[[NSAttributedString alloc] initWithString:@""]]) {
```
This is slightly inefficient as every time we perform this comparison we're doing a heap allocation just to create an empty string to compare against.  Would be better to cache it somewhere - it could even be static and shard across all instances. 
Or, better, can you could just use `lastMoveDescription.length == 0`?

joemuers commented on Matchismo/Matchismo/Model/CardMatchingGame.mm:
```diff
@@ -114,25 +131,43 @@ - (void)chooseCardAtIndex: (NSUInteger)index {
           [self.chosenCards insertObject:card atIndex:0];
-          self.lastCardFlipDescription = [NSString stringWithFormat:@"No match! %d point penalty!",
-                                          MISMATCH_PENALTY];
+          NSMutableAttributedString *descriptionString = [[[NSAttributedString alloc]
```
This whole method is way too big so it's hard to see what's going on.
If you decompose it into a bunch of smaller private methods it'll be much more readable.

joemuers commented on Matchismo/Matchismo/Model/SetCard.h:
```diff
@@ -0,0 +1,28 @@
+  setCardShadingCount
+} SetCardShading;
+
+@property (nonatomic) NSUInteger noOfShapes;
```
Preferable not to use abbreviations like this in property names, `numberOfShapes` is better.

joemuers commented on Matchismo/Matchismo/Model/SetCard.h:
```diff
@@ -0,0 +1,28 @@
+@property (nonatomic) NSUInteger noOfShapes;
+@property (nonatomic) NSString *shape;
+@property (nonatomic) SetCardShading shading;
+@property (nonatomic) UIColor *color;
```
Again here - try to keep UI stuff out of model objects, and try to avoid exposing mutable types publicly.

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+  
+  switch (self.shading) {
+    case setCardShadingSolid:
+      att[NSStrokeWidthAttributeName] = @-5;
```
Bit weird that we're using -5 as a stroke width.  Is that intentional?

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+#include "SetCard.h"
+
+@interface SetCard()
+- (BOOL)allSameNoOfShapes:(NSArray<SetCard *> *)otherCards;
```
Again here - better not to use the "no" abbreviation.  A bit of verbosity is fine if it helps readability.

joemuers replied:
Though I'd add that these method names don't quite make sense. Or at least - they're not clear.
What exactly are "allSameNumberOfShapes" and "allDifferentNumberOfShapes" checking for?

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+
+- (BOOL)allDifferentNoOfShapes:(NSArray<SetCard *> *)otherCards {
+  NSMutableSet *set = [[[NSSet alloc] init] mutableCopy];
+  [set addObject:@(self.noOfShapes)];
```
We're just adding one object to the set here, and then never changing it.
Looks like a mistake?

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+}
+
+- (BOOL)allDifferentNoOfShapes:(NSArray<SetCard *> *)otherCards {
+  NSMutableSet *set = [[[NSSet alloc] init] mutableCopy];
```
you can just use `[[NSMutableSet alloc] init]` or `[NSMutableSet set]`;

joemuers replied:
Also, NSSet/MSMutableSet are genericised in the same way as NSArray and NSDictionary, so preferable to do `NSMutableSet<NSNumber *> *`.

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+- (BOOL)allDifferentShape:(NSArray<SetCard *> *)otherCards {
+  NSMutableSet *set = [[[NSSet alloc] init] mutableCopy];
+  [set addObject:self.shape];
+  for (SetCard *card in otherCards) {
```
Same mistake with the set.
Were you assuming that `[set containsObject:card.shape]` also adds to the set as a side effect? (it doesn't).

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+  return YES;
+}
+
+- (MatchResult) match:(NSArray *)otherCards {
```
Minor - wrong spacing here (between return type & method name).

joemuers commented on Matchismo/Matchismo/Model/SetCard.mm:
```diff
@@ -0,0 +1,165 @@
+                                  || [self allDifferentShading:otherCards];
+  const auto colorCondition = [self allSameColor:otherCards]
+                                || [self allDifferentColor:otherCards];
+  result.score = (noOfShapesCondition && shapeCondition
```
Bit strange for a couple of reasons:
- we're assigning a value to `result.score`, then immediately checking it afterwards and re-assigning a different value afterwards (better to just use a local `BOOL` variable I think).
- We're checking `result.score` as if it's a boolean, but it's of type `int`.

joemuers commented on Matchismo/Matchismo/Model/SetCardDeck.mm:
```diff
@@ -0,0 +1,36 @@
+
+- (instancetype)initWithShapes:(NSArray *)shapes Colors:(NSArray *)colors {
+  if (self = [super init]) {
+    if (shapes.count != 3 || colors.count != 3) return nil;
```
Objective-C doesn't have a direct equivalent of Swift's failable initialisers, so returning `nil` like this from an init method isn't the right approach.
Firstly it'll give you a memory leak (as the caller has already invoked `alloc`).  But also it could also cause really unexpected behaviour as someone calling `auto myVar = [[SetCardDeck alloc] initWithShapes:shape colors:colors]` will never imagine that the result could be `nil`. 
If it's a _really_ bad error then you could throw an exception or something (and hence crash the app), but in most cases it would probably be better to use a factory method instead.  So something like:
```
+ (nullable instancetype)setCardDeckWithShapes:(NSArray *)shapes Colors:(NSArray *)colors {
  if (shapes.count != 3 || colors.count != 3) {
     return nil;
  }
  return [[self alloc] initWithShapes:shapes colors:colors];
}
```

joemuers commented on Matchismo/Matchismo/Model/SetCardDeck.h:
```diff
@@ -0,0 +1,14 @@
+#import "Deck.h"
+
+@interface SetCardDeck : Deck
+
```
We generally try to avoid object-oriented approaches and prefer composition and protocol-oriented approaches wherever possible (just an FYI - I get that this is the approach that the exercise pushes you into).

But in the cases where we do use inheritance, we always try to use the `NS_DESIGNATED_INITIALIZER` macro so that we get as much support from the compiler as possible (you'll also need to combine it with `NS_UNAVAILABLE` in some cases to disallow initialisers from parent classes).

joemuers reviewed (changes requested):
Great work - looks like you're really getting the hang of Obj-C in general.  Slightly weird exercise this, as obviously the sample code that they give you isn't in the Lightricks "style" and also it's a good few years old so some of the paradigms they're guiding you towards are kind of outdated.

Quite a few comments I've added are as much for guidance than because something is "wrong" as such.  Hopefully that's obvious from the content. 

In general, if you weren't already aware, going all-in on an object-oriented approach is generally to be avoided (i.e. prefer composition over inheritance, try to favour protocols rather than base classes) but I get that this is what the exercise is basically forcing you to do.

Also we never use storyboards or xibs (they are awful), but again I get that this is what the exercise is giving you.

Only other point to make for now is that we often try to split PRs over multiple commits, as this can make the reviews easier to parse (i.e. so you can digest it in chunks), and also because it means the commit history will be much more descriptive about what changes were made & why, if you see what I mean.

But anyway - good job!

Have a go at either actioning the comments, or replying to any you disagree with or want more clarity about, and then maybe we can F2F if there's anything you want to discuss/ask about.

otto-lightricks committed:
Modify exercise 3

Refactor UI elements out of model classes into VCs.
Split long functions into smaller ones.
Replace constants with enums.
Rename class to explicitly mark as abstract.
